src/java.base/share/classes/jdk/internal/foreign/LayoutRecordMapper.java

```diff
@@ -0,0 +1,641 @@
+                                            // (MemorySegment, GroupLayout, long offset, long count)
+                                            mh = MethodHandles.insertArguments(mh, 4, componentMapper);
+                                            // (MemorySegment, GroupLayout, long offset, long count) ->
+                                            // (MemorySegment, GroupLayout, long offset)
+                                            mh = MethodHandles.insertArguments(mh, 3, info.sequences().getFirst().elementCount());
+                                            // (MemorySegment, GroupLayout, long offset) ->
+                                            // (MemorySegment, long offset)
+                                            mh = MethodHandles.insertArguments(mh, 1, gl);
+                                            // (MemorySegment, long offset) -> (MemorySegment)Record[]
+                                            mh = MethodHandles.insertArguments(mh, 1, byteOffset);
+                                            // (MemorySegment, long offset)Record[] -> (MemorySegment)componentType
+                                            yield MethodHandles.explicitCastArguments(mh, MethodType.methodType(cl.component().getType(), MemorySegment.class));
+                                        } catch (NoSuchMethodException | IllegalAccessException e) {
+                                            throw new RuntimeException(e);
+                                        }
+                                    }
+                                    case SequenceLayout __ -> {
+                                        throw new InternalError("Should not reach here");
+                                    }
+                                    case PaddingLayout __ -> throw fail(cl);
+                                }
+                            } catch (NoSuchMethodException | IllegalAccessException e) {
+                                throw new InternalError(e);
+                            }
+                        }
+                        case PaddingLayout __ -> throw fail(cl); // Ignore
+                    };
+                })
+                .filter(Objects::nonNull) // Remove ignored items
+                .toArray(MethodHandle[]::new);
+
+        try {
+            var ctor = PUBLIC_LOOKUP.findConstructor(type, MethodType.methodType(void.class, ctorParameterTypes));
+            for (int i = 0; i < handles.length; i++) {
+                // Insert the respective handler for the constructor
+                ctor = MethodHandles.filterArguments(ctor, i, handles[i]);
+            }
+
+            var mt = MethodType.methodType(type, MemorySegment.class);
+            // Fold the many identical MemorySegment arguments into a single argument
+            ctor = MethodHandles.permuteArguments(ctor, mt, IntStream.range(0, handles.length)
```

Comment:
What is the special meaning of using `IntStream`? Why not `new int[handles.length]`?